### DISCUSSION TITLE
Migrating from epoll_wait to liburing

### DISCUSSION THREAD

**Discussion by gamedev8** [axboe/liburing#520] [Q&A] [answered]
Hello - I'm trying to migrate one of our projects (web client only) over to liburing, and just trying to understand something. When you use `epoll_wait(...)` you provide an out param that will be filled with the "ready list" of `epoll_event` structures. You would check them and deal with the events as needed. For example:

```
// just example code!

// create and submit a file descriptor watcher to epoll
struct epoll_event ev;
ev.events = EPOLLIN | EPOLLOUT | EPOLLHUP | EPOLLERR;
ev.data.fd = fd;

epoll_ctl(epfd, EPOLL_CTL_ADD, fd, &ev);

struct epoll_event events[8];
while (isRunning) {
   int n = epoll_wait(epfd, events, 8, -1);

   for (int i = 0; i < n; i += 1) {
      // check "events[i].events" to see what happened
   }
}
```

In liburing I see you can prep with `epoll_event` structures, but how can you get the ready `epoll_event` structures when you start querying the completion queue?
```
// just example code!
// watch a file descriptor

struct epoll_event ev;
ev.events = EPOLLIN | EPOLLOUT | EPOLLHUP | EPOLLERR;
ev.data.fd = fd;

io_uring_sqe *sqe = io_uring_get_sqe(ring)
io_uring_prep_epoll_ctl(sqe, epfd, fd, EPOLL_CTL_ADD, &ev);
io_uring_submit(ring);

io_uring_wait_cqe(&ring, cqes);
for (io_uring_cqe* cqe: cqes) {
   // ???
}
```

**Comment by axboe** [comment, marked as answer]
Really not much difference if you just want to use it as a direct epoll replacement, you would just use ```io_uring_for_each_cqe()```, for example, to iterate available CQEs in the CQ ring after wait returns. Then don't forget to use ```io_uring_cq_advance()``` when done looping to consume the number of entries that you handled.

**Reply by gamedev8** [reply]
ok thanks

**Reply by gamedev8** [reply]
Wait a second, so let's say I'm trying to check for `EPOLLHUP`. How would I do that?

**Reply by axboe** [reply]
When you iterate the cqes, then ```cqe->res``` will be the result for each operation. This is part of the liburing/io_uring fundamentals, would suggest you read a primer on how it works to get more familiar with it.
